### PR TITLE
Add asset package creation

### PR DESCRIPTION
**Pull request by dwillist**
Signed-off-by: dwillist <dthornton@vmware.com>

## Summary
Implement `asset-cache` creation as specified by #1055

This adds the experimental `create-asset-cache` command that works as follows:
`pack create-asset-cache <image-name> --buildpack <buildpack locator> --publish(optional) --pull-policy(optional)`

Additionally it adds support for using asset caches when
- Running `pack build ` with the `--asset-cache` flag
- Adding asset caches to builders.

~~(side UI bug, local `buildpack` require a `file://` prefix...~~

## Documentation
- Should this change be documented?
    - [x] Yes,  (whole feature to be documented upon completion) 
    - [ ] No

## Related

Implements first part of #1055

### CODE REVIEW

**Comment by dfreilich** [comment]
This is more of a meta-point, but what do you think about calling the command `pack asset-cache create`, instead of `create-asset-cache`? Do we expect any further `asset-cache` commands? 

As we're close to finished with #597, I'm hesitant to add more commands which don't fit in with the new paradigm we're trying to follow.

File: internal/blob/downloader.go
Comment (aemengo) [inline internal/blob/downloader.go:45]: 🤔 I'm really curious why we're removing the pointer receiver, through this whole file.

File: build.go
Comment (aemengo) [inline build.go:749]: Should this be exported? It's not referenced outside the package.

File: internal/blob/downloader.go
Comment (dwillist) [reply internal/blob/downloader.go:45]: Ya this is so that we can set `downloader.blobOptions` & `downloader.validationSha256` in the download method & prevent it mutating the underlying `downloader` object

File: acceptance/acceptance_test.go
Comment (dwillist) [inline acceptance/acceptance_test.go:2256]: There is definitely some simplification here, but I just want to push off doing a refactor of these tests because there will be many additions when testing asset-cache usage

File: create_asset_cache.go
Comment (dwillist) [inline create_asset_cache.go:56]: Note the cache poisoning & inability to exit using `ctrl+c` is not new behavior

File: create_asset_cache.go
Comment (aemengo) [inline create_asset_cache.go:84]: I'm struggling to see why we need `r.index` here. Our final return type is a mapping so preserving order doesn't matter.
Are we expecting duplicates to be returned from the channel with a given `r.sha256`?

Comment (dwillist) [reply create_asset_cache.go:84]: > Are we expecting duplicates to be returned from the channel with a given r.sha256?

Exactly. Two buildpacks share an asset, it has the same `r.sha256` but different metadata. If we don't always pick the same metadata we are missing a little bit of reproducibility (though this should be rare...).

If we remove this constraint and fiddle with the timings [here](https://github.com/buildpacks/pack/pull/1085/files#diff-27a7971b2d46c736fecf53ce767289b1b4121623503b68948654b51f8333a98aR96-R101), we can get failures.

**Comment by ekcasey** [comment]
@dwillist I wonder if this feature should only be supported for buildpack API `0.7`+ (which doesn't exist yet). Older buildpack APIs don't support assets in `buildpack.toml`.

https://github.com/buildpacks/spec/issues/190

File: internal/blob/downloader.go
Comment (jromero) [reply internal/blob/downloader.go:45]: This sounds like a smell. What is the purpose of associating `blobOptions` to the downloader when it's really only an option for a single operation? Same can be said for `validationSha256`. It looks like it's adding unnecessary state to a service (`Downloader`) which now isn't reusable because of its state....

I would propose looking for alternatives to providing optional arguments.

For example, `DownloadOptions` can be a `func (*downloadOptions)` and `downloadOptions` be an unexported struct with the two fields you added to `Downloader`. 

Then instead of

```
	for _, option := range options {
		option(&d)
	}
```

do something like:

```
        do := &downloadOptions{}
	for _, option := range options {
		option(&do)
	}
```

**Comment by jromero** [comment]
Blocked by pending spec changes. If we want to get this feature in, then we'd want to gate it behind experimental.

**Comment by dwillist** [comment]
### PR overview
As things stand I think this is ready for PR review, I'll be updating this comment with additional acceptance testing materials for folks if the want to use them.

#### Whats implemented
Note that all of the below are hidden behind `experimental` (even the new `pack build ...` flags)
- Add a `pack asset-package create` command
  - this command takes list of buildpacks and creates an asset cache image as an `image` or `tarball`.
  - TODO: gate the creation of packages based on a `buildpack API` version. We cannot really test this until we have a lifecycle that supports the new API.
- Add the ability to add `assets` to a builder by specifying them in the `builder.toml` configuration file. (Add an example file here)
  - TODO gate the creation of builders with assets on the `platform API` version. Again we need to have the lifecycle release to get the actual platform API we should be gating on. 
- Add the ability to add `assets` to a pack build 
  - again gate this ability based on the platform API that will be used by pack & the underlying builder.

### Whats left to do.
Once there is a lifecycle release that solidifies the `platform API` and `buildpack API` needed to use assets we can gate the usage of asset caches around the following support table:

| pack used to build supports platform api >= 0.8 | pack used to create builder supports platform api >= 0.8 | lifecycle supports platform api >= 0.8 | conclusion                                                                                                                                                                                                                                      |
|-------------------------------------------------|----------------------------------------------------------|----------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| T                                               | T                                                        | T                                      | CNB_ASSETS can be set in build env! and buildpacks can use it                                                                                                                                                                                   |
| T                                               | T                                                        | F                                      | CNB_ASSETS cannot be used as lifecycle does not support API Do not allow assets to be added to builders like this.                                                                                                                              |
| T                                               | F                                                        | T                                      | CNB_ASSETS will not be set on the builder :( and assets were not allowed on the  builder when it was created. But we should support adding assets using: "pack build ... --asset-package" at build time by adding CNB_ASSETS to the environment |
| T                                               | F                                                        | F                                      | CNB_ASSETS cannot be used as lifecycle does not support API...                                                                                                                                                                                  |
| F                                               | T                                                        | T                                      | Cannot use assets as pack used to build does not support platform api 0.8                                                                                                                                                                       |
| F                                               | T                                                        | F                                      | CNB_ASSETS cannot be used as lifecycle does not support API...                                                                                                                                                                                  |
| F                                               | F                                                        | T                                      | Cannot use assets as pack used to build does not support platform api 0.8                                                                                                                                                                       |
| F                                               | F                                                        | F                                      | CNB_ASSETS cannot be used as lifecycle does not support API...                                                                                                                                                                                  |

**Comment by dwillist** [comment]
@dfreilich any chance you could take a look at this? I know its pretty large but would be nice to try and get this into the next release if possible.

File: acceptance/invoke/pack_fixtures.go
Comment (dfreilich) [inline acceptance/invoke/pack_fixtures.go:92]: What do these do, together?

File: builder/config_reader.go
Comment (dfreilich) [inline builder/config_reader.go:39]: Do we need a dedicated `AssetImageOrURI`, or would `dist.ImageOrUri` work?

File: client.go
Comment (dfreilich) [inline client.go:187]: I find this line pretty odd, given that `client.assetFetcher` is `nil` in this block.

File: config/os.go
Comment (dfreilich) [inline config/os.go:10]: This is Docker OS? ImageOS? or host machine OS? It's a bit unclear from the file alone

File: create_asset_package.go
Comment (dfreilich) [inline create_asset_package.go:49]: Overall, this func reads really well. It's decomposed super nicely. Great job!

File: create_asset_package.go
Comment (dfreilich) [inline create_asset_package.go:125]: This comment doesn't really add much

File: internal/asset/fetcher.go
Comment (dfreilich) [inline internal/asset/fetcher.go:18]:
```suggestion
```
nit

File: internal/asset/locator.go
Comment (dfreilich) [inline internal/asset/locator.go:34]: Is there some way we can combine this with the existing `locator` searching mechanism, or is it totally unfit/unintuitive to do so?

File: builder/config_reader.go
Comment (dwillist) [reply builder/config_reader.go:39]: Ya I made a new object here because `dist.ImageOrURI` has a `buildpackURI` object inside of it. Which reads awkwardly when using assets.

File: client.go
Comment (dwillist) [reply client.go:187]: Could you elaborate? I'm having trouble understanding whats off here.

File: internal/asset/locator.go
Comment (dwillist) [reply internal/asset/locator.go:34]: Definitly could but it would likely take quite a bit of work here to pull out all the `buildpack` locator specific logic. Was thinking this would be part of the larger refactoring story.

File: acceptance/invoke/pack_fixtures.go
Comment (dwillist) [reply acceptance/invoke/pack_fixtures.go:92]: They delete all the files contents so that the order of operations in the `TemplateFile` function is 
1) read the file's contents into memory
2) delete them
3) re-write out the newly templated contents to the file. 

Note this is only done on files that have been moved into a `tmp` directory for testing

File: client.go
Comment (dfreilich) [reply client.go:187]: Ah never mind, I thought you were passing `client.assetFetcher` into the function, not `client.imageFetcher`

**Review by dfreilich — APPROVED** [review]
I've checked it over (not quite as in depth as I would like) and it definitely seems to be good. 

To be honest, I don't have the time allocated to the project right now to give it the in depth review it really needs (on the scale of 4+ hours, I would assume), and its current iteration as one massive PR isn't the most conducive for it being a comprehensively reviewed PR, unfortunately. At the same time, since it's gated with an `experimental` flag, I'm ok merging it in, assuming that there will be further refactors and simplifications/combinations (for instance, seeing whether we can combine some of the primitives which seem to be redefined, such as the separate Fetchers, locators, etc). 

Given that (and how long it's been waiting), I'll give it a approval now, but would definitely appreciate smaller PRs in the future 😅

**Comment by dfreilich** [comment]
@dwillist Looks like there's some issues with acceptance tests which needs to be fixed before merging

**Pull request closed without merging**
